llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -10661,7 +10704,16 @@ template <typename BVTy, typename ResTy, typename... Args>
 ResTy BoUpSLP::processBuildVector(const TreeEntry *E, Args &...Params) {
   assert(E->State == TreeEntry::NeedToGather && "Expected gather node.");
   unsigned VF = E->getVectorFactor();
-
+  BVTy ShuffleBuilder(Params...);
+  // FIXME: Only full gathering is implemented for non-power-of-2 nodes at the
+  // moment.
+  if (E->isNonPowOf2Vec()) {
+    Value *BV = ShuffleBuilder.gather(E->Scalars);
+    SmallVector<int> Mask(VF, PoisonMaskElem);
+    std::iota(Mask.begin(), Mask.end(), 0);
+    ShuffleBuilder.add(BV, Mask);
+    return ShuffleBuilder.finalize(std::nullopt);
+  }
```

Comment:
Again, why do you need special processing for non-power-of-2? It would be good, if you could try to adjust this whole function to support it. May help in future for the major non-power-of-2 functionality.

Reply:
It's not needed in the latest version, removed, thanks!